File: lib/src/test/java/io/cqrs/core/event/InMemoryEventRepository.java
package io.cqrs.core.event;

import io.cqrs.core.identifiers.EntityId;

import javax.annotation.Nonnull;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

public class InMemoryEventRepository implements EventRepository {

    private List<EventEnvelope<? extends Event, ? extends EntityId<?>>> eventStore = new ArrayList<>();

    @Nonnull
    @Override
    public <T, EI extends EntityId<T>> List<EventEnvelope<? extends Event, ? extends EntityId<?>>> listAllForEntity(@Nonnull final EI entityId) {
        return eventStore.stream()
                .filter(eventEnvelope -> eventEnvelope.getEventCoreData().getEntityId().equals(entityId))
                .collect(Collectors.toList());
    }

    @Override
    public void write(@Nonnull final List<EventEnvelope<? extends Event, ? extends EntityId<?>>> events) {
        eventStore.addAll(events);
    }


//    @Nonnull
//    @Override
//    public <E extends Event, EI extends EntityId> List<EventEnvelope<E, EI>> listAllForEntity(@Nonnull final EI entityId) {
//        return (List<EventEnvelope<E, EI>>) eventStore.stream()
//                .filter(eventEnvelope -> eventEnvelope.getEventCoreData().getEntityId().equals(entityId));
//    }
//
//    @Override
//    public <V extends Event, EI extends EntityId> void write(@Nonnull final List<EventEnvelope<V, EI>> events) {
//        eventStore.addAll(events);
//    }


}

File: kotlin-demo/src/main/kotlin/io/cqrs/kt/librarymanager/db/InMemoryEventRepository.kt
package io.cqrs.kt.librarymanager.db

import io.cqrs.core.event.Event
import io.cqrs.core.event.EventEnvelope
import io.cqrs.core.event.EventRepository
import io.cqrs.core.identifiers.EntityId
import org.slf4j.LoggerFactory

/**
 * An in-memory implementation of the EventRepository. Data will not be persisted between startups, of course,
 * and this is merely used for example purposes.
 */
class InMemoryEventRepository: EventRepository {
    private val journal = mutableListOf<EventEnvelope<out Event, out EntityId<*>>>()

    override fun <EI : EntityId<*>> listAllForEntity(entityId: EI): List<EventEnvelope<out Event, out EntityId<*>>> =
        journal
            .filter { eventEnvelope -> eventEnvelope.eventCoreData.entityId == entityId }
            .sortedBy { eventEnvelope -> eventEnvelope.eventCoreData.revision }

    override fun write(events: List<EventEnvelope<out Event, out EntityId<*>>>) {
        journal.addAll(events)
        log.debug("Successfully recorded ${events.size} events. Event Store now contains ${journal.size} events")
    }

    fun reset() {
        journal.removeAll { it.eventCoreData.revision > 2 }
    }

    companion object {
        private val log = LoggerFactory.getLogger(InMemoryEventRepository::class.java)
    }

}
